src/workerd/api/actor.h

```diff
@@ -60,23 +61,66 @@ class DurableObjectRpcStub : public Fetcher, public jsg::NamedIntercept {
       JSG_REQUIRE(!isRegisteredName(name), Error, "cannot call registered method '", name, "'");
 
       // Named intercept is enabled, this means we won't default to legacy behavior.
-      // TODO(soon): The server side PR will fill this function with the RPC call and other stuff,
-      //             this is just a temporary placeholder for testing.
-      return jsg::JsValue(js.wrapReturningFunction(js.v8Context(),
-          [](jsg::Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) {
-            return js.str("foo"_kj);
+      return jsg::JsValue(js.wrapReturningFunction(js.v8Context(), [this, methodName=kj::str(name)]
+          (jsg::Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) {
+            kj::Vector<jsg::JsValue> argv(args.Length());
+            for (int n = 0; n < args.Length(); n++) {
+              argv.add(jsg::JsValue(args[n]));
+            }
+            auto& ioContext = IoContext::current();
+            // TODO(now): Just a reminder for the future, rename this operation name.
+            auto worker = getClient(ioContext, kj::none, "rpcTest"_kjc);
+            auto event = kj::heap<api::GetJsRpcTargetCustomEventImpl>(9);
+
+            rpc::JsRpcTarget::Client client = event->getCap();
+            auto builder = client.callRequest();
+            builder.setMethodName(methodName);
+
+            // If we have arguments, serialize them.
+            if (argv.size() > 0) {
+              auto ser = jsg::serializeV8Rpc(js, js.arr(argv.asPtr()));
+              builder.initSerializedArgs().setV8Serialized(kj::mv(ser));
+            }
+
+            auto callResult = builder.send();
+            auto customEventResult = worker->customEvent(kj::mv(event))
+                .attach(kj::mv(worker), kj::mv(client));
+
+            auto jsPromise = ioContext.awaitIo(js, kj::mv(callResult),
+                [&ioContext, custom = kj::mv(customEventResult)]
+                (jsg::Lock& js, capnp::Response<rpc::JsRpcTarget::CallResults> rpcResult) mutable
+                    -> jsg::Promise<jsg::Value> {
+
+              // TODO(now): Can we skip awaiting and checking the custom event result? Maybe just
+              // attach the promise to callResult above?
+              return ioContext.awaitIo(js, kj::mv(custom), [rpcResult=kj::mv(rpcResult)]
```

Comment:
I'd suggest joining `callResult` and `custom` into a single KJ promise, so that `awaitIo()` only needs to be used once. The isolate lock has to be acquired each time `awaitIo` completes, so as currently written, a lock is acquired after the first `awaitIo` and then immediately released to begin the second `awaitIo`. This redundant lock is expensive and could be avoided by awaiting both promises in KJ land without entering the isolate.

Reply:
https://github.com/cloudflare/workerd/compare/7a33e501a03d136d4c42eaec57cf98ebfd887934..f3f824fa18ba81ebe18d0f8f0dac69eaa2e1fa7a

Had to do a bit of a work around to get the promise results back but I think this should do.